Source/DafnyServer/CounterExampleGeneration/DafnyModel.cs

```diff
@@ -458,34 +527,64 @@ public class DafnyModel {
           utfCode = ((Model.Integer)fCharToInt.OptEval(elt)).AsInt();
           return "'" + char.ConvertFromUtf32(utfCode) + "'";
         }
-        return "?#" + GetShortElementId(elt);
+        return GetUnspecifiedValue(elt, Type.Char);
       }
       if (fType.OptEval(elt) == fReal.GetConstant()) {
         if (fU2Real.OptEval(elt) != null) {
           return CanonicalName(fU2Real.OptEval(elt));
         }
-        return "?#" + GetShortElementId(elt);
+        return GetUnspecifiedValue(elt, Type.Real);
       }
       if (fType.OptEval(elt) == fBool.GetConstant()) {
         if (fU2Bool.OptEval(elt) != null) {
           return CanonicalName(fU2Bool.OptEval(elt));
         }
-        return "?#" + GetShortElementId(elt);
+        return GetUnspecifiedValue(elt, Type.Bool);
       }
       if (fType.OptEval(elt) == fInt.GetConstant()) {
         if (fU2Int.OptEval(elt) != null) {
           return CanonicalName(fU2Int.OptEval(elt));
         }
-        return "?#" + GetShortElementId(elt);
+        return GetUnspecifiedValue(elt, Type.Int);
       }
       return "";
     }
 
-    private int GetShortElementId(Model.Element element) {
-      if (!shortElementIds.ContainsKey(element.Id)) {
-        shortElementIds[element.Id] = shortElementIds.Count;
+    private string GetUnspecifiedValue(Model.Element element, Type primitiveType) {
+      if (reservedValuesMap.ContainsKey(element)) {
+        return reservedValuesMap[element];
+      }
+      if (primitiveType == Type.Bool) {
+        if (!reservedValues[Type.Bool].Contains("true")) {
+          reservedValues[Type.Bool].Add("true");
+          reservedValuesMap[element] = "true";
+        } else {
+          reservedValuesMap[element] = "false";
+        }
+        return reservedValuesMap[element];
+      }
+
+      var i = 0;
+      if (primitiveType == Type.Char) {
```

Comment:
For me it's hard to follow what these sections are intended to do without studying the code line by line. I suggest extracting these parts into separate methods to make their intent clear. That helps anyone else getting easier a big picture as well as checking if the small method really does what it's meant to do.

Reply:
I have split this method in three and added a docstring to each, so the code should be more readable